text/0000-dyno.md

```diff
@@ -0,0 +1,276 @@
+
+Lets examine the changes to `Error` required to make this work:
+
+```rust
+pub mod error {
+    pub trait Error: Debug + Provider {
+        ...
+        fn provide_context<'a>(&'a self, _req: &mut Demand<'a>) {}
+    }
+
+    impl<T: Error> Provider for T {
+        fn provide<'a>(&'a self, req: &mut Demand<'a>) {
+            self.provide_context(req);
+        }
+    }
+
+    impl dyn Error + '_ {
+        ...
+        pub fn get_context_ref<T: ?Sized + 'static>(&self) -> Option<&T> {
+            crate::any::request_ref(self)
+        }
+    }
+}
+```
+
+`get_context_ref` is added as a method on `Error` trait objects (`Error` is also likely to support similar methods for values and possibly other types, but I will elide those details), it simply calls `any::request_ref` (we'll discuss this function below). But where does the context data come from? If a concrete error type supports backtraces, then it must override the `provide_context` method when implementing `Error` (by default, the method does nothing, i.e., no data is provided, so `get_context_ref` will always returns `None`). `provide_context` is used in the blanket implementation of `Provider` for `Error` types, in other words `Provider::provide` is delegated to `Error::provide_context`.
+
+Note that by adding `provide_context` with a default empty implementation and the blanket `impl` of `Provider`, these changes to `Error` are backwards compatible. However, this pattern is only possible because `Provider` and `Error` will be defined in the same crate. For third party users, users will implement `Provider::provide` directly, in the usual way.
+
+In `provide_context`, an error type provides access to data via a `Demand` object, e.g., `req.provide_ref::<Backtrace>(&self.backtrace)`. The type of the reference passed to `provide_ref` is important here (and we encourage users to use explicit types with turbofish syntax even though it is not necessary, this might even be possible to enforce using a lint). When a user calls `get_context_ref`, the requested type must match the type of an argument to `provide_ref`, e.g., the type of `&self.backtrace` is `&Backtrace`, so a call to `get_context_ref::<Backtrace>()` will return a reference to `self.backtrace`. An implementer can make multiple calls to `provide_ref` to provide multiple data with different types.
+
+Note that `Demand` has methods for providing values as well as references, and for providing more complex types. These will be covered in the next section.
+
+The important additions to `any` are
+
+```rust
+pub trait Provider {
+    fn provide<'a>(&'a self, req: &mut Demand<'a>);
+}
+
+pub fn request_value<'a, T: 'static>(provider: &'a dyn Provider) -> Option<T> { ... }
+pub fn request_ref<'a, T: ?Sized + 'static>(provider: &'a dyn Provider) -> Option<&'a T> { ... }
+
+pub struct Demand<'a>(...);
```

Comment:
Would it be possible to add another lifetime specifying for how long the provided value must live?

Reply:
It isn't (I think). The way things work is that the lifetime is named in the trait but not the type tag (which must be bounded by `'static`). But that means that the `T` in `Value<T>` needs the `'static` bound. You can write a more specific tag for something like `Foo<'a>`, but you can't extend the general `Value` tag to include a generic lifetime.

Reply:
Right, forgot about `TypeId` requiring `'static`.